npm build: create addons directory that loads from lib. Closes #3780

iamdustan opened the pull request:
This generates the following files:

![screen shot 2015-04-29 at 4 10 22 pm](https://cloud.githubusercontent.com/assets/227879/7400357/4e9f6840-ee8a-11e4-8655-ee9449fc870a.png)

zpao commented:
I actually think we should match up with the fields on `React.addons` now. I don't think we'll be able to fully automate this. eg, `react/addons/createFragment` needs to look like `module.exports = require('../lib/ReactFragment').create;`

iamdustan commented:
Ah. So looking at `build/npm-react/lib/ReactWithAddons.js`, it looks like the following are the not-straight export ones:

```
CSSTransitionGroup: ReactCSSTransitionGroup,
PureRenderMixin: ReactComponentWithPureRenderMixin,
TransitionGroup: ReactTransitionGroup,
batchedUpdates: ReactUpdates.batchedUpdates
createFragment: ReactFragment.create,
Perf: ReactDefaultPerf,
TestUtils: ReactTestUtils,
```

Dang. Three items in and I realized it would’ve been easier to have the default ones done. What is easier for your team for maintenance going forward? Just manually writing in the npm build source the mapping or look into using `esprima-fb` for the transform.

Should be pretty straightforward.

Visit all `React.addons` keys. Create a new file with the key name with the value.

zpao commented:
I liked @spicyj's idea of just adding the files directly into the repo. Though I guess he said "build/npm-react" which isn't right and I think he miswrote. We can do it before the build process starts.

If we create an addons folder in https://github.com/facebook/react/tree/master/npm-react with the right files and the right content, we don't have to do anything fancy. It's more prone to falling out of date but perhaps we can do some simple testing.

spicyj commented:
Yeah, my bad – I meant just npm-react/.

iamdustan commented:
...and updated. Is this what you were looking for?

Also, I added a deprecation notice for `require("react/addons");` in a similar style to other `deprecated` notifications I found in the code. I think that should only be called once in userland since require calls are cached, so I didn’t wrap it.

spicyj commented:
I think these all need to be `../lib` not `./lib`. Were you able to test these at all? If you run grunt npm-react:pack then you should be able to install from build/react.tgz.

zpao commented:
That's it (except for the paths as Ben mentioned) :)

A couple things I'm not sure about:
1. The deprecation notice. It might end up being that we really do want to deprecate this but I'm not entirely sure yet.
2. What to do about `__DEV__`. Right now `React.addons.TestUtils` doesn't exist at all in prod. Same with `Perf`. Now maybe it doesn't matter so much since people would be requiring explicitly but it is a slight difference.

spicyj commented:
I'm personally up for deprecating react/addons now (though I'd probably do a console.warn) and I don't think the `__DEV__` difference is that important here.

iamdustan commented:
Ah. When I first searched for how other deprecations were being used I ended up in the JSXTransformer. In the React source I’m seeing the `warning` module used. Should I use that instead?

https://github.com/facebook/react/search?l=javascript&q=%22is+deprecated%22&type=Code&utf8=%E2%9C%93

I have successfully installed from the .tgz file. Doing a quick, manual sanity check right now.

iamdustan commented:
Looks legit to me:

![screen shot 2015-04-29 at 5 30 53 pm](https://cloud.githubusercontent.com/assets/227879/7401906/b27d2da6-ee95-11e4-8d14-15c6712525f4.png)

iamdustan commented:
>  2. What to do about __DEV__. Right now React.addons.TestUtils doesn't exist at all in prod. Same with Perf. Now maybe it doesn't matter so much since people would be requiring explicitly but it is a slight difference.

I really like explicit change. It is very attractive to be able to move my team to `var TestUtils = require('react/addons/TestUtils)` and disallow requiring the entire `react/addons` in app code.

spicyj commented on npm-react/addons.js:
```diff
@@ -1 +1,8 @@
+warning(
+  false,
+  'require("react/addons") is deprecated. ' +
+  'Access using require("react/addons/{addon}" instead.'
```
Missing paren here. Otherwise looks good to me – mind squashing your commits?

iamdustan commented:
Done.

spicyj commented on npm-react/addons.js:
```diff
@@ -1 +1,8 @@
+warning(
+  false,
+  'require("react/addons") is deprecated. ' +
+  'Access using require("react/addons/{addon})" instead.'
```
`")` not `)"` :)

iamdustan commented:
Doh! I swear I’m paying attention. :innocent:

spicyj commented:
Thank you!

zpao commented:
Thanks Dustan!

iamdustan commented:
Np! I’ll take these 30 minute `[Good First Bug]`s any day!
